reference/reference.go

```diff
@@ -0,0 +1,280 @@
+	tail = s
+	if p.err != nil || len(s) == 0 || s[0] != ':' {
+		return
+	}
+	tag = tagRegexp.FindString(s[1:])
+	if len(tag) > 0 {
+		tail = s[len(tag)+1:]
+	}
+	return
+}
+
+// parseDigest parses the digest part of a reference.
+func (p *parser) parseDigest(s string) (dgst digest.Digest, tail string) {
+	tail = s
+	if p.err != nil || len(s) == 0 || s[0] != '@' {
+		return
+	}
+	dgst, p.err = digest.ParseDigest(s[1:])
+	if p.err != nil {
+		p.err = ErrReferenceInvalidFormat
+		return
+	}
+	tail = s[len(dgst)+1:]
+	return
+}
+
+// parse starts the parsing of the reference.
+func (p *parser) parse(s string) (Reference, error) {
+	if p.err != nil {
+		return nil, p.err
+	}
+
+	repository, s := p.parseRepository(s)
+	tag, s := p.parseTag(s)
+	dgst, s := p.parseDigest(s)
+	if len(s) > 0 {
+		p.err = ErrReferenceInvalidFormat
+	}
+	if p.err != nil {
+		return nil, p.err
+	}
+
+	if tag != "" && dgst != "" {
+		return NewTagDigestReference(repository, tag, digest.Digest(dgst)), nil
+	}
+	if dgst != "" {
+		return NewDigestReference(repository, digest.Digest(dgst)), nil
+	}
+	if tag != "" {
+		return NewTagReference(repository, tag), nil
+	}
+	return NewRepositoryReference(repository), nil
+}
+
+// Parse parses s and returns a syntactically valid Reference.
+// If an error was encountered it is returned, along with a nil Reference.
+func Parse(s string) (Reference, error) {
+	return new(parser).parse(s)
+}
+
+// DigestReference represents a reference of the form `repository@sha256:deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef`.
+// Implements the Reference interface.
+type DigestReference struct {
+	repository string
+	digest     digest.Digest
+}
+
+// Repository returns the repository part.
+func (r DigestReference) Repository() string { return r.repository }
+
+// String returns the full string reference.
+func (r DigestReference) String() string {
+	return r.repository + "@" + string(r.digest)
+}
+
+// NewDigestReference returns an initialized DigestReference.
+func NewDigestReference(repo string, digest digest.Digest) DigestReference {
+	return DigestReference{
+		repo,
+		digest,
+	}
+}
+
+// TagReference represents a reference of the form `repository:tag`.
+// Implements the Reference interface.
+type TagReference struct {
+	repository string
+	tag        string
+}
+
+// Repository returns the repository part.
+func (r TagReference) Repository() string { return r.repository }
+
+// String returns the full string reference.
+func (r TagReference) String() string {
+	return r.repository + ":" + r.tag
+}
+
+// NewTagReference returns an initialized TagReference.
+func NewTagReference(repo string, tag string) TagReference {
+	return TagReference{
+		repo,
+		tag,
+	}
+}
+
+// TagDigestReference represents a reference of the form `repository:tag@sha256:deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef`
+// with both tags and digests. Implements the Reference interface.
+type TagDigestReference struct {
```

Comment:
Rather than having a `TagDigestReference`, `DigestReference` should just have an optional tag set. In nearly all cases, `TagDigestReference` will be treated as a `DigestReference` except that it is annotated with a source tag.